### PR TITLE
2019: Rewrite MCI using acornjs (depends on #116)

### LINKED ISSUE #24: Metacircular interpreter's parser uses AY17/18 Source Lanugage specs
### Problem
- The parser that is generated using `jison` uses grammar that works for the previous implementation for Source
- The rules would have to be changed
  - [ ] `var` must be changed to `const` and `let`
  - [ ] Arrow functions must be added
  - [ ] Anonymous functions must be removed (?)

### PR DESCRIPTION
**Pull request by thomastanck**
New version of #115 

Resolves #24.

### CODE REVIEW

**Comment by thomastanck** [comment]
Agreed, this change is definitely for next semester.

I think thats why @openorclose suggested to split off #116 as that one is not a breaking change.

**Comment by martin-henz** [comment]
116 would disallow:

(function(x){return x+1;})(1);

We can't do that, because mission grading depends on it. Can't change this
during this semester. Need to wait for next.

-martin
http://www.comp.nus.edu.sg/~henz

**Comment by thomastanck** [comment]
Should I disable that change for now, to be re-enabled next semester?

**Comment by martin-henz** [comment]
Yes no changes please at this stage, unless it's serious bugs. Tx.

-martin

**Comment by martin-henz** [comment]
This will break when objects are removed. Let's have a CP3108 project on an object-free MCI, and shelve this PR.

**Comment by thomastanck** [comment]
We should probably still merge this in first so we can remove the old parser. The only thing left to change are the transformation rules for turning the estree spec compliant parse tree into a CS1101S compliant one.

**Comment by martin-henz** [comment]
#116 is merged now. Feel free to merge this one. We can take a closer look later. I'll try to get a CP3108 team on this.

File: scripts/build.sh [hunk 1]
```diff
@@ -3,4 +3,4 @@
 # This is required because it is not possible to both inlcude .js
```
Comment (openorclose) [inline scripts/build.sh]: Time to remove these obsolete comments?

File: src/stdlib/parser.ts [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+  })}],
+  ["ConditionalExpression", (node: es.Node) => {
+    node = <es.ConditionalExpression> node
+    return ({
+      tag: "conditional_expression",
+      predicate: transform(node.test),
+      consequent: transform(node.consequent),
+      alternative: transform(node.alternate)
+  })}],
+  ["FunctionExpression", (node: es.Node) => {
```
Comment (openorclose) [inline src/stdlib/parser.ts]: Time to remove FunctionExpression

File: src/stdlib/parser.ts [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    if (node.key.type === 'Literal') {
+      return [node.key.value, transform(node.value)];
+    } else if (node.key.type === 'Identifier') {
+      return [node.key.name, transform(node.value)];
+    } else {
+      unreachable()
+      throw new ParseError('Invalid property key type')
+    }
+  }],
+  ["EmptyStatement", (node: es.Node) => {
```
Comment (openorclose) [inline src/stdlib/parser.ts]: And empty statements

File: src/stdlib/parser.ts [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    node = <es.BreakStatement> node
+    return ({
+      tag: "break_statement"
+  })}],
+  ["ContinueStatement", (node: es.Node) => {
+    node = <es.ContinueStatement> node
+    return ({
+      tag: "continue_statement"
+  })}],
+  ["ThisExpression", (node: es.Node) => {
```
Comment (openorclose) [inline src/stdlib/parser.ts]: Remove this as well.

File: scripts/build.sh (hunk 1 above)
Comment (thomastanck) [reply scripts/build.sh]: Removed the build script entirely as it's now useless

File: src/stdlib/parser.ts (hunk 2 above)
Comment (thomastanck) [reply src/stdlib/parser.ts]: Removed all the transformers you listed as well.

**Pull request merged**